src/server/dflycmd.cc

```diff
@@ -59,6 +65,110 @@ struct TransactionGuard {
   Transaction* t;
 };
 
+// Buffered single-shard journal streamer that listens for journal changes with a
+// journal listener and writes them to a destination sink in a separate fiber.
+class JournalStreamer {
+ public:
+  JournalStreamer(io::Sink* dest, journal::Journal* journal, Context* cntx)
+      : cntx_{cntx}, journal_{journal}, waker_{}, dest_{dest}, write_fb_{}, writer_{},
+        journal_cb_id_{0} {
+  }
+
+  // Register journal listener and start writer in fiber.
+  void Start();
+
+  // Must be called on context cancellation for unblocking
+  // and manual cleanup.
+  void Cancel();
+
+ private:
+  // Writer fiber that steals buffer contents and writes them to dest.
+  void WriterFb();
+
+  // Common condition for both sides to check that consumer is blocked
+  // because it buffered too much and waits for the writer to finish writing.
+  bool IsStalled() const;
+
+ private:
+  Context* cntx_;
+  journal::Journal* journal_;
+
+  unsigned buffered_ = 0;                 // how many entries are buffered
+  ::util::fibers_ext::EventCount waker_;  // two sided waker
+
+  io::Sink* dest_;
+  Fiber write_fb_;
+  JournalWriter writer_;
+
+  uint32_t journal_cb_id_;
+};
+
+void JournalStreamer::Start() {
+  write_fb_ = Fiber(&JournalStreamer::WriterFb, this);
+
+  journal_cb_id_ = journal_->RegisterOnChange([this](const journal::Entry& entry) {
+    writer_.Write(entry);
+    buffered_++;
+
+    // Stall and wait for writer to finish if too much data is accumulated.
+    // Otherwise, notify writer if it is waiting for data.
+    if (IsStalled()) {
+      // TODO: If 0->stalled cannot happen at once, then this wake is redundand.
+      // NOTE: Before removing, don't forget about this when switching to byte estimation.
+      waker_.notify();
+      waker_.await([this]() { return !IsStalled() || cntx_->IsCancelled(); });
```

Comment:
I would write it differently (lines 111-122):
```
if (buffered_++ == 0) {
  waker_.notify();  // notify the producer in case it's waiting for a buffer to read.
} else {
  // this won't have any overhead if the cond evaluates to true.
  waker_.await([this]() { return !IsStalled() || cntx_->IsCancelled(); });
}
```